Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java

```diff
@@ -1342,12 +1340,12 @@ private void updateInformationFromServer(DocumentEvent e) {
         final int eventOffset = e.getOffset();
         final int deltaLength = e.getNewLength() - e.getOldLength();
         for (PluginNavigationRegion region : regions) {
-          if (region.offset <= eventOffset && eventOffset <= region.offset + region.length) {
-            region.length += deltaLength;
-          }
-          else if (region.offset >= eventOffset) {
+          if (eventOffset <= region.offset) {
             region.offset += deltaLength;
           }
+          else if (region.offset < eventOffset && eventOffset <= region.offset + region.length) {
+            region.length += deltaLength;
```

Comment:
Ok, I'll update the code and add tests. For navigation - delete if ranges intersect. For highlighting: update if do not intersect or if DocumentEvent range is inside region. Delete if intersect like in my previous examples.